BugSplatDotNetStandard/BugSplat.cs

```diff
@@ -124,22 +128,80 @@ public async Task<HttpResponseMessage> Post(Exception ex, ExceptionPostOptions o
-        public async Task<HttpResponseMessage> Post(FileInfo minidumpFileInfo, MinidumpPostOptions options = null)
+        public async Task<HttpResponseMessage> Post(FileInfo minidumpFileInfo, MinidumpPostOptions options = null, string md5 = "")
         {
             ThrowIfArgumentIsNull(minidumpFileInfo, "minidumpFileInfo");
 
+            options = options ?? new MinidumpPostOptions();
+
+            var crashUploadResponse = await getCrashUploadUrl(minidumpFileInfo);
+
+            ThrowIfHttpRequestFailed(crashUploadResponse);
+
+            var presignedUrl = await deserializeCrashUploadUrl(crashUploadResponse);
+            var uploadFileResponse = await uploadFileToPresignedURL(presignedUrl, minidumpFileInfo);
+
+            ThrowIfHttpRequestFailed(uploadFileResponse);
+
+            var commitS3CrashResponse = await commitS3CrashUpload(presignedUrl.ToString(), options, md5);
+
+            ThrowIfHttpRequestFailed(commitS3CrashResponse);
+
+            return commitS3CrashResponse;
+        }
+
+
+        private async Task<HttpResponseMessage> getCrashUploadUrl(FileInfo fileInfo)
+        {
+            string path = $"{baseUrl}/api/getCrashUploadUrl";
+            string route = $"{path}?database={database}&appName={application}&appVersion={version}&crashPostSize={fileInfo.Length}";
+
             using (var httpClient = new HttpClient())
             {
-                options = options ?? new MinidumpPostOptions();
+                return await httpClient.GetAsync(route);
+            }
+        }
+
+        private async Task<Uri> deserializeCrashUploadUrl(HttpResponseMessage response)
+        {
+            var json = await response.Content.ReadAsStringAsync();
+            try
+            {
+                var awsResponse = JsonConvert.DeserializeObject<AWSResponse>(json);
+                return new Uri(awsResponse.Url);
+            }
+            catch
+            {
+                throw new JsonException("Failed to deserialize crash upload url");
+            }
+        }
+
+        private async Task<HttpResponseMessage> uploadFileToPresignedURL(Uri uri, FileInfo file)
+        {
+            using (var body = new MultipartFormDataContent())
+            {
+                body.Add(new ByteArrayContent(File.ReadAllBytes(file.FullName)), "minidump");
+                using (var httpClient = new HttpClient())
+                {
+                    httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/octet-stream"));
+                    return await httpClient.PutAsync(uri, body);
+                }
+            }
+        }
+
+        private async Task<HttpResponseMessage> commitS3CrashUpload(string s3Key, MinidumpPostOptions options, string md5 = "")
+        {
+            using (var httpClient = new HttpClient())
+            {
+                var route = $"{baseUrl}/api/commitS3CrashUpload";
```

Comment:
route is technically just the /api/commitS3CrashUpload part, a better name here would be Url or Uri